In [177]:
# Purpose of this notebook is to compare the results of several randomly selected track sets to result of the brain tracks selected for gwas_1
# Author: Sophie Sigfstead

In [178]:
keys = ['a','b','c','d','e','f', 'g','h','i','j']
sd = 1.0
base_path = "./gwas_1_leading_SNPs_by_track_random_results/gwas_1_leading_SNPs_by_track_random_{key}/sd={sd}/agg_results"
combined_path = "./gwas_1_leading_SNPs_by_track_random_results/gwas_1_leading_SNPs_by_track_random_{key}_combined_result/sd={sd}/combined_result.csv"
file_names = [
    "coding_snps_comparison.csv",
    "full_overlap_df.csv",
    "matching_loci_stats.csv",
    "non_coding_comparison.csv",
    "combined_result.csv"
]

In [179]:
import os
import pandas as pd
output_path = f"./aggregated_results/sd={sd}/"  # Directory to save aggregated results

# Create output directory if it doesn't exist
os.makedirs(output_path, exist_ok=True)

aggregated_data = {file_name: [] for file_name in file_names}

# Helper function to load files and append to aggregated data
def load_and_append(file_path, file_name):
    if os.path.exists(file_path):
        try:
            df = pd.read_csv(file_path)
            aggregated_data[file_name].append(df)
        except Exception as e:
            print(f"Failed to read {file_path}: {e}")
    else:
        print(f"File not found: {file_path}")

# Iterate over each key and each file type
for key in keys:
    for file_name in file_names:
        # Determine file path for each file type
        if file_name == "combined_result.csv":
            file_path = f"./gwas_1_leading_SNPs_by_track_random_results/gwas_1_leading_SNPs_by_track_random_{key}_combined_result/sd={sd}/{file_name}"
        else:
            file_path = f"./gwas_1_leading_SNPs_by_track_random_results/gwas_1_leading_SNPs_by_track_random_{key}/sd={sd}/agg_results/{file_name}"
        
        # Attempt to load and append the data
        load_and_append(file_path, file_name)

# Concatenate and save each aggregated file type
for file_name, data_list in aggregated_data.items():
    if data_list:  # Only proceed if there's data to aggregate
        aggregated_df = pd.concat(data_list, ignore_index=True)
        aggregated_df.to_csv(os.path.join(output_path, file_name), index=False)
        print(f"Saved aggregated {file_name} to {output_path}")
    else:
        print(f"No data found for {file_name} across the specified directories.")

Saved aggregated coding_snps_comparison.csv to ./aggregated_results/sd=1.0/
Saved aggregated full_overlap_df.csv to ./aggregated_results/sd=1.0/
Saved aggregated matching_loci_stats.csv to ./aggregated_results/sd=1.0/
Saved aggregated non_coding_comparison.csv to ./aggregated_results/sd=1.0/
Saved aggregated combined_result.csv to ./aggregated_results/sd=1.0/


In [180]:
def column_stats (df):
    numeric_df = df.select_dtypes(include=['number'])
    
    # calculate required statistics and round to two decimals
    summary_df = pd.DataFrame({
        'Mean': numeric_df.mean().round(2),
        'Median': numeric_df.median().round(2),
        'Standard Deviation': numeric_df.std().round(2),
        'Max': numeric_df.max().round(2),
        'Min': numeric_df.min().round(2)
    })
    return summary_df



### Matching Loci Dataframes -  Aggregation
- in the analysis, the mean, median, and standard_dev of the matching loci for each track was computed, as was correlation between non-coding:total snps, coding:total snps found.

In [181]:
## Matching Loci Average
matching_loci_agg = pd.read_csv("./aggregated_results/sd=1.0/matching_loci_stats.csv")
matching_loci_agg

mean_matching_loci  median_matching_loci  standard_dev_matching_loci  \
0           21.441176                  22.0                    3.491906   
1           22.941176                  24.0                    3.064372   
2           22.117647                  22.0                    2.782681   
3           22.235294                  23.0                    2.892612   
4           22.352941                  22.5                    3.338321   
5           22.147059                  22.0                    2.840533   
6           21.647059                  21.5                    3.292621   
7           23.117647                  23.0                    3.505534   
8           21.294118                  21.0                    3.289371   
9           22.352941                  22.0                    2.993755   

   correlation_non_coding_to_total_snps_found  \
0                                    0.979278   
1                                    0.971210   
2                                    0.971810   
3                                    0.975983   
4                                    0.968671   
5                                    0.972059   
6                                    0.976208   
7                                    0.971177   
8                                    0.980919   
9                                    0.977920   

   correlation_coding_to_total_snps_found  
0                               -0.714041  
1                               -0.653449  
2                               -0.791247  
3                               -0.692465  
4                               -0.534491  
5                               -0.733036  
6                               -0.690994  
7                               -0.702560  
8                               -0.758800  
9                               -0.668558

In [182]:
reference_col = [24.7,25, 1.51, 0.97, -0.76]
result = column_stats(matching_loci_agg)
result["Ref"] = reference_col
result

Mean  Median  Standard Deviation  \
mean_matching_loci                          22.16   22.19                0.59   
median_matching_loci                        22.30   22.00                0.86   
standard_dev_matching_loci                   3.15    3.18                0.27   
correlation_non_coding_to_total_snps_found   0.97    0.97                0.00   
correlation_coding_to_total_snps_found      -0.69   -0.70                0.07   

                                              Max    Min    Ref  
mean_matching_loci                          23.12  21.29  24.70  
median_matching_loci                        24.00  21.00  25.00  
standard_dev_matching_loci                   3.51   2.78   1.51  
correlation_non_coding_to_total_snps_found   0.98   0.97   0.97  
correlation_coding_to_total_snps_found      -0.53  -0.79  -0.76

Important findings in above result: 
1. mean loci per track lower (n = 330) at 22.16 versus 24.7, but not significantly lower than the mean for the brain tracks. 
2. Median loci range (min to max) for random track sets was fully under the reference. 
3. Correlation seems similar as well, indicating a consistent negative relation between number of coding snps and total snps found, and a consistent positive relationship between the number of non-coding snps and number of total snps found. 
Overall, its clear that the reference set does do better, just not as much as expected.

### Full Overlap DF Aggregation

In [183]:
full_overlap_df_agg = pd.read_csv("./aggregated_results/sd=1.0/full_overlap_df.csv")
full_overlap_df_agg = full_overlap_df_agg.reset_index(drop=True).drop_duplicates()
full_overlap_df_agg = full_overlap_df_agg[full_overlap_df_agg['track_col']!="ALL_brain"]
full_overlap_df_agg = full_overlap_df_agg.sort_values(by=['matching_loci'], ascending=False)
full_overlap_df_agg['stem_cell_tissue'] = full_overlap_df_agg['tissue'].str.contains("stem cell", na=False)
full_overlap_df_agg['embryo_tissue'] = full_overlap_df_agg['tissue'].str.contains("embryo", na=False)
full_overlap_df_agg['adult_tissue'] = full_overlap_df_agg['target_labels'].str.contains("adult", na=False)
full_overlap_df_agg['DNASE'] = full_overlap_df_agg['target_labels'].str.contains("DNASE", na=False)
full_overlap_df_agg.head(15)

track_col  matching_snps  matching_loci  new_coding_snps  \
240     SAD32              4             29                6   
238    SAD665             10             29                4   
34     SAD624              7             29                7   
205    SAD399              7             28                8   
172    SAD293              6             28                5   
242    SAD534              7             28                5   
206    SAD362              6             28                7   
243    SAD394              4             28                7   
171    SAD542              7             28                6   
139     SAD30              4             27                6   
138     SAD11              5             27                6   
273     SAD44              2             27                5   
103    SAD576              3             27                7   
173     SAD43              4             26                6   
245    SAD433              3             26                8   

                                         target_labels           tissue  \
240  DNASE:iPS-NIHi7 female adult (85 years) origin...        stem cell   
238                 DNASE:testis male adult (54 years)      testis male   
34                                            DNASE:H9        stem cell   
205         DNASE:sigmoid colon female adult (51 year)    sigmoid colon   
172                      DNASE:iPS DF 6.9 male newborn        stem cell   
242                DNASE:vagina female adult (51 year)           vagina   
206          DNASE:tibial nerve female adult (51 year)            nerve   
243          DNASE:Peyer's patch male adult (37 years)  small intestine   
171  DNASE:H9 G1 phase genetically modified using s...        stem cell   
139                                   DNASE:CWRU1 male        stem cell   
138                                      DNASE:GM10248   lymphoblastoid   
273  DNASE:psoas muscle male adult (27 years) and m...     psoas muscle   
103   DNASE:limb embryo (53 days) and embryo (56 days)             limb   
173                                   DNASE:osteoblast             bone   
245   DNASE:limb embryo (58 days) and embryo (59 days)             limb   

     stem_cell_tissue  embryo_tissue  adult_tissue  DNASE  
240              True          False          True   True  
238             False          False          True   True  
34               True          False         False   True  
205             False          False          True   True  
172              True          False         False   True  
242             False          False          True   True  
206             False          False          True   True  
243             False          False          True   True  
171              True          False         False   True  
139              True          False         False   True  
138             False          False         False   True  
273             False          False          True   True  
103             False          False         False   True  
173             False          False         False   True  
245             False          False         False   True

In [184]:
full_overlap_df_agg_stats= column_stats(full_overlap_df_agg)
reference_col = [4.21,24.7,7.36]
full_overlap_df_agg_stats['Ref'] = reference_col 
full_overlap_df_agg_stats

Mean  Median  Standard Deviation  Max  Min    Ref
matching_snps     3.22     3.0                1.60   10    0   4.21
matching_loci    21.94    22.0                2.99   29   14  24.70
new_coding_snps   9.11     9.0                2.14   14    4   7.36

In [185]:
full_overlap_df_agg_stats_adult= column_stats(full_overlap_df_agg[full_overlap_df_agg['adult_tissue']])
full_overlap_df_agg_stats_adult

Mean  Median  Standard Deviation  Max  Min
matching_snps     3.49     3.0                1.76   10    1
matching_loci    20.78    20.0                3.97   29   14
new_coding_snps   9.02     9.0                2.30   14    4

In [186]:
full_overlap_df_agg_stats_embryo= column_stats(full_overlap_df_agg[full_overlap_df_agg['embryo_tissue']])
full_overlap_df_agg_stats_embryo

Mean  Median  Standard Deviation  Max  Min
matching_snps     3.64     4.0                1.22    5    2
matching_loci    22.93    24.0                2.20   26   19
new_coding_snps   8.71     8.0                1.73   12    6

In [187]:
full_overlap_df_agg_stats_stem =  column_stats(full_overlap_df_agg[full_overlap_df_agg['stem_cell_tissue']])
full_overlap_df_agg_stats_stem

Mean  Median  Standard Deviation  Max  Min
matching_snps     3.60     4.0                1.96    7    1
matching_loci    24.20    25.0                3.88   29   18
new_coding_snps   8.33     7.0                2.58   13    5

In [188]:
full_overlap_df_agg_stats_dnase =  column_stats(full_overlap_df_agg[full_overlap_df_agg['DNASE']])
full_overlap_df_agg_stats_dnase

Mean  Median  Standard Deviation  Max  Min
matching_snps     3.22     3.0                1.61   10    0
matching_loci    21.90    22.0                3.00   29   14
new_coding_snps   9.11     9.0                2.15   14    4

Important findings:

1. Stem cell tracks (mean loci = 24.2) performed best, just a bit lower than  the singular brain tracks on average (mean loci = 24.7). This was followed by embryonic tracks (mean loci = 22.93), then adult tissue tracks (mean = 20.78).
2. These tracks showed the same trend for matching snps.
3. It seemed that increased accuracy (# loci) across adult, embyro and stem cell sets resulted in less coding snps being chosen.
4. Despite some DNASE tracks scoring quite highly, it does not appear that that DNASE were more accurate than ATAC tracks.

### Combined Result Comparison
- The purpose of this analysis is to compare how sets of randomly selected tracks perform together compared to the brain set. That is, these are the results of replicating the GWAS method using a random set of tracks. 


In [189]:
import pandas
combined_result_agg = pd.read_csv("./aggregated_results/sd=1.0/combined_result.csv").sort_values(by=['num_loci_overlap'], ascending=True)
combined_result_agg 

num_snps_found  num_snps_overlap  num_coding_snps  num_loci_overlap  \
0              36                13                3                27   
1              38                14                3                28   
2              36                10                3                28   
3              36                13                3                28   
4              38                12                3                28   
5              37                14                3                28   
6              37                14                3                28   
7              35                16                3                28   
8              39                10                3                28   
9              36                14                3                28   

        p_value  
0  1.562753e-07  
1  1.518037e-07  
2  1.633208e-07  
3  1.535632e-07  
4  1.538515e-07  
5  1.433336e-07  
6  1.393949e-07  
7  1.110236e-07  
8  1.502377e-07  
9  1.331000e-07

The brain track set found 38 loci including 27 overlapping loci, 6 common snps and 3 coding snps. These results are quite similar. I am wondering if the regions that these loci exist in are quite large and therefore active sites housing many important SNPs. The autocomplete protocol says that "all significant hits within 1 Mb from the SNP were ignored, yeilding 2Mb windows. I know this is the correct interpretation as we reproduced the study exactly with the 0.0 threshold. 

However, in our report, we found that the non-matching, but overlapping snps had a largest distance of 558284 base pairs away from the original snp.
So, I'm thinking it would be worthwhile to run these analyses with smaller windows (e.g. 500k base pairs) to compare how our method fairs against random track sets. 

### Coding SNPs Comparison - Aggregation
- Compare the number of coding snps found by singular tracks 

In [190]:
coding_snps_comparison_agg = pd.read_csv("./aggregated_results/sd=1.0/coding_snps_comparison.csv")
coding_snps_comparison_agg.sort_values(by='track_coding_snps', ascending=False).head()
coding_snps_comparison_agg['stem_cell_tissue'] = coding_snps_comparison_agg['tissue'].str.contains("stem cell", na=False)
coding_snps_comparison_agg['embryo_tissue'] =coding_snps_comparison_agg['tissue'].str.contains("embryo", na=False)
coding_snps_comparison_agg['adult_tissue'] = coding_snps_comparison_agg['target_labels'].str.contains("adult", na=False)
coding_snps_comparison_agg_stats = column_stats(coding_snps_comparison_agg)
coding_snps_comparison_agg_stats["ref"] = [7.5, 3, 4.5]
coding_snps_comparison_agg_stats

Mean  Median  Standard Deviation  \
track_coding_snps                        9.07     9.0                2.14   
intersection_with_all_track_coding_snps  3.00     3.0                0.00   
number_of_unique_snps                    6.07     6.0                2.14   

                                         Max  Min  ref  
track_coding_snps                         14    4  7.5  
intersection_with_all_track_coding_snps    3    3  3.0  
number_of_unique_snps                     11    1  4.5

In [191]:
coding_snps_comparison_agg_stats_adult = column_stats(coding_snps_comparison_agg[coding_snps_comparison_agg['adult_tissue']])
coding_snps_comparison_agg_stats_adult

Mean  Median  Standard Deviation  \
track_coding_snps                        8.88     9.0                2.16   
intersection_with_all_track_coding_snps  3.00     3.0                0.00   
number_of_unique_snps                    5.88     6.0                2.16   

                                         Max  Min  
track_coding_snps                         14    4  
intersection_with_all_track_coding_snps    3    3  
number_of_unique_snps                     11    1

In [192]:
coding_snps_comparison_agg_stats_stem = column_stats(coding_snps_comparison_agg[coding_snps_comparison_agg['stem_cell_tissue']])
coding_snps_comparison_agg_stats_stem

Mean  Median  Standard Deviation  \
track_coding_snps                        8.29     7.0                2.45   
intersection_with_all_track_coding_snps  3.00     3.0                0.00   
number_of_unique_snps                    5.29     4.0                2.45   

                                         Max  Min  
track_coding_snps                         13    5  
intersection_with_all_track_coding_snps    3    3  
number_of_unique_snps                     10    2

In [193]:
coding_snps_comparison_agg_stats_embryo = column_stats(coding_snps_comparison_agg[coding_snps_comparison_agg['embryo_tissue']])
coding_snps_comparison_agg_stats_embryo

Mean  Median  Standard Deviation  \
track_coding_snps                        8.75     8.0                1.65   
intersection_with_all_track_coding_snps  3.00     3.0                0.00   
number_of_unique_snps                    5.75     5.0                1.65   

                                         Max  Min  
track_coding_snps                         12    6  
intersection_with_all_track_coding_snps    3    3  
number_of_unique_snps                      9    3

Important findings: 
1. The 3 new coding snps found via our method were always recovered. 
2. The more accurate tissue sets seemed to have less coding snps. More work can be done to link up coding snps and accuracy of the track set. 

### Non-Coding SNPs - Aggregation 

- Compare the number of non-coding snps found by singular tracks 

In [194]:
non_coding_snps_comparison_agg = pd.read_csv("./aggregated_results/sd=1.0/non_coding_comparison.csv")
non_coding_snps_comparison_agg.head()

track_col  num_snps_total  num_non_coding  num_coding  \
0    SAD408              27              16          11   
1    SAD281              30              22           8   
2     SAD99              33              22          11   
3    SAD526              41              31          10   
4    SAD384              40              31           9   

   num_non_coding_intersect  \
0                         4   
1                         7   
2                         8   
3                         9   
4                         7   

                non_coding_intersect_with_all_tracks  \
0  {'rs200949', 'rs10512249', 'rs4801157', 'rs102...   
1  {'rs200949', 'rs1028750', 'rs2399576', 'rs1439...   
2  {'rs200949', 'rs114707880', 'rs2399576', 'rs49...   
3  {'rs200949', 'rs215807', 'rs1028750', 'rs38962...   
4  {'rs200949', 'rs1028750', 'rs1950834', 'rs1439...   

   num_non_coding_intersect_original  \
0                                  3   
1                                  1   
2                                  3   
3                                  4   
4                                  4   

                  non_coding_intersect_with_original  num_overlap_loci  \
0            {'rs200949', 'rs10512249', 'rs4801157'}                16   
1                                       {'rs200949'}                19   
2             {'rs200949', 'rs3872822', 'rs4801157'}                22   
3  {'rs3896224', 'rs10512249', 'rs200949', 'rs195...                27   
4  {'rs200949', 'rs10752206', 'rs4801157', 'rs195...                26   

                                        overlap_loci  num_missing_loci  \
0  {'rs200949', 'rs3896224', 'rs10512249', 'rs175...                15   
1  {'rs200949', 'rs35538328', 'rs3896224', 'rs107...                13   
2  {'rs3896224', 'rs10752206', 'rs35538328', 'rs1...                10   
3  {'rs3896224', 'rs10752206', 'rs111333124', 'rs...                 5   
4  {'rs3896224', 'rs10752206', 'rs71367544', 'rs1...                 6   

                                        missing_loci  \
0  {'rs10818400', 'rs35538328', 'rs10789340', 'rs...   
1  {'rs10818400', 'rs13174273', 'rs4404022', 'rs3...   
2  {'rs111333124', 'rs10818400', 'rs1504746', 'rs...   
3  {'rs4404022', 'rs9262120', 'rs71367544', 'rs75...   
4  {'rs10818400', 'rs1504746', 'rs9262120', 'rs38...   

                                       target_labels                tissue  
0                                      DNASE:OCI-LY7       b-cell lymphoma  
1  DNASE:mammary epithelial cell female adult (18...         mammary gland  
2  DNASE:dermis blood vessel endothelial cell fem...  blood vessel to skin  
3          DNASE:mesendoderm originated from H1-hESC             stem cell  
4         DNASE:muscle of back male embryo (96 days)           back muscle

In [195]:

non_coding_snps_comparison_agg['stem_cell_tissue'] = non_coding_snps_comparison_agg['tissue'].str.contains("stem cell", na=False)
non_coding_snps_comparison_agg['embryo_tissue'] =non_coding_snps_comparison_agg['tissue'].str.contains("embryo", na=False)
non_coding_snps_comparison_agg['adult_tissue'] = non_coding_snps_comparison_agg['target_labels'].str.contains("adult", na=False)
non_coding_snps_comparison_agg_stats = column_stats(non_coding_snps_comparison_agg)
non_coding_snps_comparison_agg_stats["REF"] = [36, 33,3, None, 3, None, None]
non_coding_snps_comparison_agg_stats

Mean  Median  Standard Deviation  Max  \
num_snps_total                     37.04    38.0                5.14   48   
num_non_coding                     27.97    28.0                6.80   41   
num_coding                          9.07     9.0                2.14   14   
num_non_coding_intersect            7.66     7.0                2.81   16   
num_non_coding_intersect_original   3.27     3.0                1.63   10   
num_overlap_loci                   24.07    24.0                3.38   32   
num_missing_loci                    8.07     8.0                3.05   16   

                                   Min   REF  
num_snps_total                      26  36.0  
num_non_coding                      13  33.0  
num_coding                           4   3.0  
num_non_coding_intersect             1   NaN  
num_non_coding_intersect_original    0   3.0  
num_overlap_loci                    16   NaN  
num_missing_loci                     0   NaN

In [196]:
non_coding_snps_comparison_agg_stats_adult = column_stats(non_coding_snps_comparison_agg[non_coding_snps_comparison_agg['adult_tissue']])
non_coding_snps_comparison_agg_stats_adult

Mean  Median  Standard Deviation  Max  Min
num_snps_total                     35.75    35.5                5.93   48   27
num_non_coding                     26.88    26.5                7.63   41   13
num_coding                          8.88     9.0                2.16   14    4
num_non_coding_intersect            8.34     7.5                2.83   15    4
num_non_coding_intersect_original   3.56     3.0                1.72   10    1
num_overlap_loci                   22.95    22.0                4.03   31   16
num_missing_loci                    8.95    10.0                3.76   16    2

In [197]:
non_coding_snps_comparison_agg_stats_stem = column_stats(non_coding_snps_comparison_agg[non_coding_snps_comparison_agg['stem_cell_tissue']])
non_coding_snps_comparison_agg_stats_stem

Mean  Median  Standard Deviation  Max  Min
num_snps_total                     40.29    41.0                6.12   47   30
num_non_coding                     32.00    34.0                8.33   41   17
num_coding                          8.29     7.0                2.45   13    5
num_non_coding_intersect            9.38    10.0                3.07   15    5
num_non_coding_intersect_original   3.95     4.0                2.20    7    1
num_overlap_loci                   27.05    27.0                4.33   32   20
num_missing_loci                    5.48     5.0                4.14   12    0

In [198]:
non_coding_snps_comparison_agg_stats_embryo = column_stats(non_coding_snps_comparison_agg[non_coding_snps_comparison_agg['embryo_tissue']])
non_coding_snps_comparison_agg_stats_embryo

Mean  Median  Standard Deviation  Max  Min
num_snps_total                     38.19    38.5                3.04   44   32
num_non_coding                     29.44    30.5                4.27   36   20
num_coding                          8.75     8.0                1.65   12    6
num_non_coding_intersect            9.19     9.5                2.40   12    4
num_non_coding_intersect_original   3.56     4.0                1.21    5    2
num_overlap_loci                   24.94    26.0                2.11   28   21
num_missing_loci                    7.44     7.0                1.67   10    4

- Stem cell tracks produced the greatest number of SNPs, followed by embryo and adult. 
- Stem cell track had the greatest intersection with non-coding original snps, followed by embryo and adult. 

Potential ideas: 
- Try our analysis with loci windows of 1Mb instead of 2Mb to see if this validates our method as being more precise. Further, visualize size of window to accuracy for brain set vs random set and see if there is a point of significantly greater performance in brain set.
- Try our analysis with regular window size but only brain embryo tracks to see if its more precise.
